### PR TITLE
Edit pass on update considerations

### LINKED ISSUE #33170: Update considerations for C# needs refresh
This article probably needs a fair amount of rework now. I first wrote this to address specific feedback where customers were choosing a newer language version while targeting an older TFM (typically .NET Standard 2.0). That's not a supported scenario.

Today, customers choose a language version / TFM as a pair. That should change the focus to be more about asking what changes a developer can make that are:

- compatible API changes
- binary breaking, but source compatible API changes
- source breaking API changes

I haven't thought it through, but my first impression is that a source breaking, but binary compatible API change can't exist. 

Those changes feel like scope creep for this PR. Do you want to make the edits, or open an issue and I'll address it after break?

_Originally posted by @BillWagner in https://github.com/dotnet/docs/pull/33111#discussion_r1054711438_

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 7025d71b-df42-b8ad-3e31-20a480a5d652
* Version Independent ID: a05ae06d-91e1-d1f8-0194-e7830f5c880e
* Content: [Version and update considerations for C# developers](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/version-update-considerations)
* Content Source: [docs/csharp/whats-new/version-update-considerations.md](https://github.com/dotnet/docs/blob/main/docs/csharp/whats-new/version-update-considerations.md)
* Product: **dotnet-csharp**
* Technology: **csharp-whats-new**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**
      

---
[Associated WorkItem - 87514](https://dev.azure.com/msft-skilling/Content/_workitems/edit/87514)

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #33170

Make the language match the terms used in the runtime breaking changes docs.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/whats-new/version-update-considerations.md](https://github.com/dotnet/docs/blob/cd735024bb8b886d3309e79387c298e8147832bd/docs/csharp/whats-new/version-update-considerations.md) | [Version and update considerations for C# developers](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/version-update-considerations?branch=pr-en-us-35973) |

### CODE REVIEW

File: docs/csharp/whats-new/version-update-considerations.md [hunk 1]
```diff
@@ -2,60 +2,68 @@
...
 
-More care may be required when you adopt new language features in a library. You may be creating a new library with features found in the latest version and need to ensure apps built using previous versions of the compiler can use it. Or you may be upgrading an existing library and many of your users may not have upgraded versions yet. As you make decisions on adopting new features, you'll need to consider two variations of compatibility: source-compatible and binary-compatible.
+The language version used to compile your app typically matches the runtime TFM referenced in your project. For more information on changing the default language version, see the article on [configure your language version](../language-reference/configure-language-version.md). This default behavior ensures maximum compatibility.
```
Comment (gewarren) [inline docs/csharp/whats-new/version-update-considerations.md:12]:
```suggestion
The language version used to compile your app typically matches the runtime target framework moniker (TFM) referenced in your project. For more information on changing the default language version, see the article titled [configure your language version](../language-reference/configure-language-version.md). This default behavior ensures maximum compatibility.
```

**Review by gewarren — COMMENTED** [review]
@billwagner Unless it is different for C#, I don't think these definitions are quite right. A binary incompatible change is much more impactful than a source incompatible change. A binary BC means you have to recompile. A source BC means *if* you recompile, then you must make some source code changes.

**Comment by BillWagner** [comment]
Thanks @gewarren 

Just made updates. The definitions should match, it's just that the LDM has a focus for making sure code that has compiled before still compiles.

File: docs/csharp/whats-new/version-update-considerations.md [hunk 2]
```diff
@@ -2,60 +2,68 @@
...
+
+## Impact of breaking changes
+
+When you add a *binary breaking change* to your library, you force all projects that use your library to recompile. However, none of the source code in those projects needs to change. As a result, the impact of the breaking change is reasonably small for each project.
+
+When you make a *source breaking change* to your library, you require all projects to make source changes in order to use your new library. If the necessary change requires new language features, you force those projects to upgrade to the same language version and TFM you're now using. You've required more work for your users, and possibly forced them to upgrade as well.
+
+The impact of any breaking change you make depends on the number of projects that have a dependency on your library. If your library is used internally by a few applications, you can react to any breaking changes in all impacted projects. However, if your library is publicly downloaded, you should evaluate the potential impact and consider alternatives. You may add new APIs that parallel existing APIs. You may consider parallel builds for different TFMs. You can also consider multi-targeting.
```
Comment (gewarren) [inline docs/csharp/whats-new/version-update-considerations.md:69]:
```suggestion
The impact of any breaking change you make depends on the number of projects that have a dependency on your library. If your library is used internally by a few applications, you can react to any breaking changes in all impacted projects. However, if your library is publicly downloaded, you should evaluate the potential impact and consider alternatives:

- You might add new APIs that parallel existing APIs.
- You might consider parallel builds for different TFMs.
- You might consider multi-targeting.
```

**Review by gewarren — APPROVED** [review]

File: docs/csharp/whats-new/version-update-considerations.md [hunk 3]
```diff
@@ -2,60 +2,68 @@
...
+The language version used to compile your app typically matches the runtime target framework moniker (TFM) referenced in your project. For more information on changing the default language version, see the article titled [configure your language version](../language-reference/configure-language-version.md). This default behavior ensures maximum compatibility.
 
-## Binary-compatible changes
+When *breaking changes* are introduced, they're classified as:
 
-Changes to your library are **binary-compatible** when your updated library can be used without rebuilding applications and libraries that use it. Dependent assemblies are not required to be rebuilt, nor are any source code changes required.
+- *Binary breaking change*: A binary breaking change means that your application or library has different behavior, including possibly crashing, when launched using a new runtime. You must recompile your app to incorporate these changes. The existing binary won't function correctly.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md:16]: For the language, behavioral changes are a subset of a source breaking change. We define that has your source means something different now. I added a note to clarify that.

File: docs/csharp/whats-new/version-update-considerations.md [hunk 4]
```diff
@@ -2,60 +2,68 @@
...
+Because of the goal to avoid breaking changes by the C# language team and runtime team, updating your application is typically a matter of updating the TFM and rebuilding the app. However, for libraries that are distributed publicly, you should carefully evaluate your policy for supported TFMs and supported language versions. You may be creating a new library with features found in the latest version and need to ensure apps built using previous versions of the compiler can use it. Or you may be upgrading an existing library and many of your users may not have upgraded versions yet.
 
-## Incompatible changes
+## Introducing breaking changes in your libraries
 
-If a change is neither **source-compatible** nor **binary-compatible**, source code changes along with recompilation are required in dependent libraries and applications.
+When you adopt new language features in your library's public API, you should evaluate if adopting the feature introduces either a binary or source breaking change for the users of your library. Any changes to your internal implementation that don't appear in the `public` or `protected` interfaces are always compatible.
```
[the remark this answers is not in the document]
Comment (gewarren) [reply docs/csharp/whats-new/version-update-considerations.md:25]: InternalsVisibleTo - sorry

**Pull request merged**